nextcord/cache.py

```diff
@@ -0,0 +1,149 @@
+    ) -> None:
+        self.maxsize: Union[int, float] = maxsize or float("inf")
+        self.ttl: Optional[Union[int, float]] = ttl
+        self.timelink: dict[Hashable, float] = {}
+        self.working: bool = False
+
+    def __getitem__(self, key):
+        """Expires items before returning getitem."""
+        self.__expire__()
+        return __get_item__(self, key)
+
+    def __setitem__(self, key, val) -> None:
+        """
+        Expires and ensures self max size is maintained before setting new item.
+        """
+        self.__expire__()
+        if len(self) >= self.maxsize:
+            for item in self:  # iter to delete first item
+                del self[item]
+                break
+        __set_item__(self, key, val)
+        self.update_time(key)
+
+    def __expire__(self) -> None:
+        """
+        Takes care of expiring keys in the timelink that have passed their ttl lifespan.
+        Uses a rolling motion for expiring, all items at the start of the timelink
+        are assumed to be the earliest created items to avoid checking every item.
+        Timelink order must be managed properly in order for expiration to work.
+        """
+        if not self.ttl or self.working:
+            return
+
+        self.working = True
+
+        expiration = now() - self.ttl
+        expired = []
+
+        for key, time in self.timelink.items():
+            if time > expiration:
+                break
+            expired.append(key)
+
+        for item in expired:
+            del self[item]
+            del self.timelink[item]
+
+        self.working = False
+
+    def get(self, key, default=None):
+        """Expire before returning from get."""
+        self.__expire__()
+        return __get__(self, key, default)
+
+    def pop(self, key, default=None):
+        """Pop from timelink before returning pop from self"""
+        __pop__(self.timelink, key, None)
+        return __pop__(self, key, default)
+
+    def get_with_update(self, key, default=None):
+        """Get an item whilst updating the timelink for the item."""
+        self.__expire__()
+        item = __get__(self, key, _MISSING)
+        if item is not _MISSING:
+            self.update_time(key)
+            return item
+        return default
+
+    def set_with_persist_ttl(self, key, val) -> bool:
+        """Setting a value without updating timelink if the key already exists."""
+        self.__expire__()
+        if len(self) >= self.maxsize:
+            for item in self:
+                del self[item]
+                break
+        __set_item__(self, key, val)
+        if key not in self.timelink:
+            self.timelink[key] = now()
+            return True
+        return False
+
+    def update_time(self, key) -> None:
+        """
+        Update the timelink for a key.
+        Creates the key anew if it does not exist in the timelink.
+        Make sure the item exists in self, could result in memory leak if used improperly!

```

Comment:
How significantly does a `if key not in self: raise KeyError` slow this down?

Reply:
It wouldn't although it's intentionally designed this way, this method is supposed to be a helper method, and the checks depends on whichever method implements it.

Reply:
Added a warning to be more implicit